### PR TITLE
Implement WebGL extensions

### PR DESCRIPTION
**Pull request by MortimerGoro**
This PR provides the base for implementing WebGL extensions. It comes with the following ones already implemented and passing all related WebGL conformance tests:

- OES_texture_float
- OES_texture_float_linear
- OES_texture_half_float
- OES_texture_half_float_linear
- OES_vertex_array_object

I'll submit other extensions like compressed textures in a separate PR to ease the review process. I included the 5 extensions in this PR because it's easier to show/review how the WebGL extension base works.

To pass all OES_texture_float_xxx tests I had to add some missing format conversions in WebGLRenderingContext.rs

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16893)

### CODE REVIEW

File: components/script/dom/webgl_extensions/ext/webglvertexarrayobjectoes.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+use dom::globalscope::GlobalScope;
+use dom::webglbuffer::WebGLBuffer;
+use dom_struct::dom_struct;
+use std::cell::Cell;
+use std::collections::HashMap;
+use webrender_traits::WebGLVertexArrayId;
+
+#[dom_struct]
+pub struct WebGLVertexArrayObjectOES {
+    reflector_: Reflector,
```
Comment (emilio) [inline components/script/dom/webgl_extensions/ext/webglvertexarrayobjectoes.rs]: Shouldn't this be a `WebGLObject` instead of a reflector, to match the IDL?

Comment (MortimerGoro) [inline components/script/dom/webgl_extensions/ext/webglvertexarrayobjectoes.rs]: ouch, yes It should be added to this one because it's a different object (not the extension itself). Shouldn't missing the WebGLObject give a compilation error?

Comment (jdm) [inline components/script/dom/webgl_extensions/ext/webglvertexarrayobjectoes.rs]: We haven't figured out a good way to check that automatically.

Comment (nox) [inline components/script/dom/webgl_extensions/ext/webglvertexarrayobjectoes.rs]: https://is.gd/Xz2rxa

Ô tristesse, ô désespoir...

File: components/script/dom/webgl_extensions/manager.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+use ref_filter_map::ref_filter_map;
+use std::cell::Ref;
+use std::collections::{HashMap, HashSet};
+use super::{ext, WebGLExtension};
+use super::wrapper::{WebGLExtensionWrapper, TypedWebGLExtensionWrapper};
+use webrender_traits::WebGLError;
+
+type GLenum = u32;
+
+const DEFAULT_DISABLED_TEX_TYPES: [GLenum; 2] = [
```
Comment (emilio) [inline components/script/dom/webgl_extensions/manager.rs]: May you add comments on all these to the rationale/spec?

File: components/script/dom/webgl_extensions/manager.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+use js::jsapi::{JSContext, JSObject};
+use js::jsval::JSVal;
+use ref_filter_map::ref_filter_map;
+use std::cell::Ref;
+use std::collections::{HashMap, HashSet};
+use super::{ext, WebGLExtension};
+use super::wrapper::{WebGLExtensionWrapper, TypedWebGLExtensionWrapper};
+use webrender_traits::WebGLError;
+
+type GLenum = u32;
```
Comment (emilio) [inline components/script/dom/webgl_extensions/manager.rs]: There's a `GLEnum` type in `gleam::gl` I believe.

File: components/script/dom/webgl_extensions/wrapper.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+use dom::webglrenderingcontext::WebGLRenderingContext;
+use heapsize::HeapSizeOf;
+use js::jsapi::JSObject;
+use std::any::Any;
+use super::{WebGLExtension, WebGLExtensionManager};
+
+/// Trait used internally by WebGLExtensionManager to store and
+// handle the different WebGL extensions in a common list.
+pub trait WebGLExtensionWrapper: JSTraceable + HeapSizeOf {
+    fn instance_or_init(&self,
```
Comment (emilio) [inline components/script/dom/webgl_extensions/wrapper.rs]: needs docs, please :)

File: components/script/dom/webgl_extensions/wrapper.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                        ctx: &WebGLRenderingContext,
+                        manager: &WebGLExtensionManager)
+                        -> NonZero<*mut JSObject>;
+    fn is_supported(&self, &WebGLExtensionManager) -> bool;
+    fn enable(&self, manager: &WebGLExtensionManager);
+    fn name(&self) -> &'static str;
+    fn as_any<'a>(&'a self) -> &'a Any;
```
Comment (emilio) [inline components/script/dom/webgl_extensions/wrapper.rs]: is this really needed? why can't you just do `self as &Any` in the relevant places?

Comment (emilio) [inline components/script/dom/webgl_extensions/wrapper.rs]: Also, I _think_ you can elide the lifetime.

File: components/script/dom/webglbuffer.rs [hunk 6]
```diff
@@ -112,6 +119,36 @@ impl WebGLBuffer {
     pub fn target(&self) -> Option<u32> {
         self.target.get()
     }
+
+    pub fn is_attached_to_vao(&self) -> bool {
+        self.vao_references.borrow().as_ref().map_or(false, |vaos| vaos.len() > 0)
```
Comment (emilio) [inline components/script/dom/webglbuffer.rs]: `!is_empty()`

File: components/script/dom/webglrenderingcontext.rs [hunk 7]
```diff
@@ -178,11 +186,13 @@ impl WebGLRenderingContext {
...
                 bound_buffer_array: MutNullableJS::new(None),
                 bound_buffer_element_array: MutNullableJS::new(None),
+                bound_attrib_buffers: DOMRefCell::new(HashMap::new()),
                 bound_renderbuffer: MutNullableJS::new(None),
                 current_program: MutNullableJS::new(None),
                 current_vertex_attrib_0: Cell::new((0f32, 0f32, 0f32, 1f32)),
                 current_scissor: Cell::new((0, 0, size.width, size.height)),
-                current_clear_color: Cell::new((0.0, 0.0, 0.0, 0.0))
+                current_clear_color: Cell::new((0.0, 0.0, 0.0, 0.0)),
+                extension_manager: WebGLExtensionManager::new()
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: It feels weird to have a `Manager` class that is a member variable (I'd have expected it to be a singleton).

Perhaps calling `WebGLExtensionManager` just `WebGLExtensions` is ok? Not a big deal though.

File: components/script/dom/webglrenderingcontext.rs [hunk 8]
```diff
@@ -296,6 +330,24 @@ impl WebGLRenderingContext {
...
+            if info.is_initialized() {
+                self.validate_filterable_texture(&texture,
+                                                 target,
+                                                 0,
+                                                 info.internal_format().unwrap_or(TexFormat::RGBA),
+                                                 info.width(),
+                                                 info.height(),
+                                                 info.data_type().unwrap_or(TexDataType::UnsignedByte));
+            }
+
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: nit: extra newline.

File: components/script/dom/webglrenderingcontext.rs [hunk 9]
```diff
@@ -329,6 +381,36 @@ impl WebGLRenderingContext {
         }
     }
 
+    // LINEAR filtering may be forbidden when using WebGL extensions (e.g. OES_TEXTURE_FLOAT_LINEAR)
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Link/quote spec?

File: components/script/dom/webglrenderingcontext.rs [hunk 10]
```diff
@@ -369,6 +451,13 @@ impl WebGLRenderingContext {
         true
     }
 
+    #[inline]
+    fn luminance(r: u8, g: u8, b: u8) -> u8 {
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Doc comments, at least describing a source for these constants (https://en.wikipedia.org/wiki/Relative_luminance may be enough)

File: components/script/dom/webglrenderingcontext.rs [hunk 11]
```diff
@@ -1019,14 +1260,21 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
 
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.14
     fn GetSupportedExtensions(&self) -> Option<Vec<DOMString>> {
-        Some(vec![])
+        self.extension_manager.init_once(||{
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: nit: space before brace, here and everywhere else.

File: components/script/dom/webglrenderingcontext.rs [hunk 12]
```diff
@@ -2885,23 +3166,32 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
 
-        self.tex_image_2d(texture, target, data_type, format,
+        self.tex_image_2d(&texture, target, data_type, format,
                           level, width, height, border, unpacking_alignment, pixels);
 
         Ok(())
     }
 
     // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.8
+    #[allow(unsafe_code)]
```
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Why is this needed? At a glance you haven't added any unsafe code.

**Review by emilio — COMMENTED** [review]
Looks pretty good to me!

I have a few comments and requests though.

File: components/script/dom/webgl_extensions/wrapper.rs (hunk 5 above)
Comment (MortimerGoro) [inline components/script/dom/webgl_extensions/wrapper.rs]: I tried to use directly the Any trait instead of adding the method but I couldn't. Here is an explanation: http://stackoverflow.com/a/33687996

Yes, the lifetime declaration was unneeded.

File: components/script/dom/webglrenderingcontext.rs (hunk 7 above)
Comment (MortimerGoro) [inline components/script/dom/webglrenderingcontext.rs]: Ok, WebGLExtensions is a good name.

**Comment by emilio** [comment]
Nice! Squash, then r=me

@bors-servo delegate+

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #249: gl.getVertexAttrib(2, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer].
  │   → assert_true: gl.getVertexAttrib(2, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer]. expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:242:9
  │ shouldBeNull@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:273:29
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:389:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #250: gl.getVertexAttrib(3, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer].
  │   → assert_true: gl.getVertexAttrib(3, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer]. expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:242:9
  │ shouldBeNull@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:273:29
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:390:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:
  │ FAIL [expected PASS] WebGL test #252: gl.getVertexAttrib(1, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer].
  │   → assert_true: gl.getVertexAttrib(1, gl.VERTEX_ATTRIB_ARRAY_BUFFER_BINDING) should be null. Was [object WebGLBuffer]. expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:242:9
  │ shouldBeNull@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:273:29
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/misc/object-deletion-behaviour.html:392:1

  ▶ OK [expected CRASH] /webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:
  │ FAIL [expected PASS] WebGL test #6: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #7: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #14: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #15: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #22: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #23: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #30: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #31: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data
  │   → assert_true: LUMINANCE_ALPHA/UNSIGNED_BYTE should maintain full precision of data expected true got false
  │ FAIL [expected PASS] WebGL test #65: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4 expected true got false
  │ FAIL [expected PASS] WebGL test #66: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1 expected true got false
  │ FAIL [expected PASS] WebGL test #69: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_4_4_4_4 expected true got false
  │ FAIL [expected PASS] WebGL test #70: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1
  │   → assert_true: UNPACK_PREMULTIPLY_ALPHA_WEBGL with RGBA/UNSIGNED_SHORT_5_5_5_1 expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:88:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:7
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:159:5
  │ runOneTest@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:453:9
  │ maybeRunTests@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:382:9
  └ ImageWrapper.prototype.generateData/img<@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/textures/tex-image-with-format-and-type.html:483:7

  ▶ ERROR [expected OK] /webgl/conformance-1.0.3/conformance/extensions/oes-texture-float-with-canvas.html
  └   → ctx.fillText is not a function

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/extensions/oes-texture-float-with-canvas.html:
  └ NOTRUN [expected PASS] Overall test

  ▶ ERROR [expected OK] /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-canvas.html
  └   → ctx.fillText is not a function

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-canvas.html:
  └ NOTRUN [expected PASS] Overall test

  ▶ TIMEOUT [expected OK] /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-video.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ DESCRIPTION: Verify texImage2D and texSubImage2D code paths taking video elements (RGBA/HALF_FLOAT_OES)
  │ PASS Successfully enabled OES_texture_half_float extension
  │ 
  │ 
  │ testing: video/mp4; codecs="avc1.42E01E, mp4a.40.2"
  └ testing: video/mp4; codecs="avc1.42E01E, mp4a.40.2"

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/extensions/oes-texture-half-float-with-video.html:
  └ NOTRUN [expected PASS] Overall test
```

**Comment by MortimerGoro** [comment]
Fixed issues:
- Update expectations for `ctx.fillText is not a function` in related tests
- Update test expectations for tex-image-with-format-and-type.html. Previously it was exiting early  because of missing format conversions.
- Fixed failing VERTEX_ATTRIB_ARRAY_BUFFER_BINDING tests (https://github.com/servo/servo/pull/16893/files#diff-7637bdaee356e2d89fec3a68089d9175R1886)

@bors-servo try
